parlai/crowdsourcing/tasks/turn_annotations_static/webapp/src/components/task_components.jsx

```diff
@@ -184,125 +49,352 @@ function SubtaskCounter({ subtaskIndex, numSubtasks }) {
+          response: null,
+        };
+        Object.keys(this.props.taskConfig.annotation_buckets.config).forEach(
+          b => {
+            answersForTurn.buckets[b] = null;
+          },
+        );
+        answersForSubtaskIndex[j] = answersForTurn;
+      }
+      answers[i] = answersForSubtaskIndex;
+    }
+    return answers;
+  }
 
-        <br />
-        <span dangerouslySetInnerHTML={{ __html: taskDescription || 'Task Description Loading' }}></span>
-        <SubtaskSubmitButton subtaskIndex={index} numSubtasks={taskData.length} onSubtaskSubmit={() => { handleSubtaskSubmit(index, setIndex, taskData.length, taskData[index], taskConfig.annotation_buckets, onSubmit); }} initialTaskData={taskData[index]}></SubtaskSubmitButton>
-      </LeftPane>
-      <RightPane>
-        <ContentPane subtaskData={taskData[index]} taskConfig={taskConfig} subtaskIndex={index} numSubtasks={taskData.length} ></ContentPane>
-      </RightPane>
-      <div style={{ clear: 'both' }}>
+  validateUserInput() {
+    console.log('Validating user input');
+    // if task is over, user should not be able to submit anything else
+    if (this.state.index >= this.state.answers.length) {
+      console.log('Task is over: ' + this.state.answers.length + ' vs ' + this.state.index);
+      return false;
+    }
+    // check that all turns have answers
+    const subtaskData = this.state.answers[this.state.index];
+    for (var i = 0; i < subtaskData.length; i++) {
+      const uttAnswer = subtaskData[i];
+      if (uttAnswer.agent_idx == 1 && uttAnswer.do_annotate) {
+        // only "bot" utterances (could be self-chat) have checkboxes
+        var atLeastOneAnswerChecked = false;
+        Object.keys(uttAnswer.buckets).forEach(b => {
+          atLeastOneAnswerChecked =
+            atLeastOneAnswerChecked || uttAnswer.buckets[b];
+        });
+        if (!atLeastOneAnswerChecked) {
+          console.log('No answers checked');
+          return false;
+        }
+      }
+    }
+    // check response fields
+    var responses = document.getElementsByName('input_response');
+    if (responses.length > 0) {
+      for (var j = 0; j < responses.length; j++) {
+        if (!validateFreetextResponse(responses[j].value, 10, 2, 2)) {
+          return false;
+        }
+      }
+    }
+    console.log('Input valid!');
+    return true;
+  }
+
+  handleSubtaskSubmit = () => {
+    const subtaskIndex = this.state.index;
+    console.log('In handleSubtaskSubmit for subtask: ' + subtaskIndex);
+    if (!this.validateUserInput()) {
+      alert('Data does not meet basic quality standards. Cannot be submitted.');
+      return;
+    }
+
+    if (subtaskIndex == this.props.taskData.length - 1) {
+      this.props.onSubmit(this.state.answers);
+    }
+    this.setState(prevState => ({index: prevState.index + 1}));
+  };
+
+  setChecked = (turnIdx, bucket, value) => {
+    var answers = this.state.answers;
+    answers[this.state.index][turnIdx].buckets[bucket] = value;
```

Comment:
The bucket value is null if it isn't checked, would be nice if you could explicitly set the answer data to true if value is true and false if value is null/false.